packages/apollo-gateway/src/__tests__/gateway/lifecycle-hooks.test.ts

```diff
@@ -0,0 +1,153 @@
+import { ApolloGateway, GatewayConfig } from '../../index';
+import * as accounts from '../__fixtures__/schemas/accounts';
+import * as books from '../__fixtures__/schemas/books';
+import * as inventory from '../__fixtures__/schemas/inventory';
+import * as product from '../__fixtures__/schemas/product';
+import * as reviews from '../__fixtures__/schemas/reviews';
+
+const services = [product, reviews, inventory, accounts, books];
+const serviceDefinitions = services.map((s, i) => ({
+  name: s.name,
+  typeDefs: s.typeDefs,
+  url: `http://localhost:${i}`,
+}));
+
+describe('lifecycle hooks', () => {
+  it('uses updateServiceDefinitions override', async () => {
+    const experimental_updateServiceDefinitions = jest.fn(
+      async (config: GatewayConfig) => {
+        return [serviceDefinitions, true];
+      },
+    );
+
+    const gateway = new ApolloGateway({
+      serviceList: serviceDefinitions,
+      experimental_updateServiceDefinitions,
+      experimental_didUpdateComputedFederationConfig: jest.fn(),
+    });
+
+    await gateway.load();
+
+    expect(experimental_updateServiceDefinitions).toBeCalled();
+  });
+
+  it('calls experimental_didFailComposition with a bad config', async done => {
+    const update = jest.fn(async (config: GatewayConfig) => {
+      return [[serviceDefinitions[0]], true];
+    });
+
+    const experimental_didFailComposition = jest.fn();
+
+    const gateway = new ApolloGateway({
+      experimental_updateServiceDefinitions: update,
+      experimental_didFailComposition,
+    });
+
+    try {
+      await gateway.load();
+    } catch (e) {
+      const callbackArgs = experimental_didFailComposition.mock.calls[0][0];
+      expect(callbackArgs.serviceList).toHaveLength(1);
+      expect(callbackArgs.errors).toMatchInlineSnapshot(`
```

Comment:
nit: should we be a bit less precise about wording on matches so that we can change that text without needing to change all of our tests? Though maybe inline snapshots make this magical!

Reply:
I actually think having the wording in the test is nice. Since snapshots can easily be changed, I don't think this'll be a problem long-term. It's also a nice thing that the error messages contain info about the schema what couldn't be composed (like the typenames)